### PR TITLE
[SYCLomatic] Support the migration of __half_raw

### LINKED ISSUE #1038: migrate __half_raw
I find __half_raw is used in some CUDA program. 

```
inline __device__ __host__ half f(unsigned short v) {
#if CUDA_VERSION >= 9000
    __half_raw h;
    h.x = v;
    return __half(h);
#else
    half h;
    h.x = v;
    return h;
#endif
}
```

### PR DESCRIPTION
**Pull request by wenhuiNi**

### CODE REVIEW

File: clang/test/dpct/half_raw.cu [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// RUN: FileCheck --input-file %T/half_raw/half_raw.dp.cpp --match-full-lines %s
+
+#include<iostream>
+#include<cuda_runtime.h>
+#include<cuda_fp16.h>
+int main(){
+  //CHECK: sycl::half one_h{sycl::bit_cast<sycl::half>(uint16_t(0x3C00))};
+  __half_raw one_h{0x3C00};
+  //CHECK: sycl::half zero_h{sycl::bit_cast<sycl::half>(uint16_t(0))};
+  __half_raw zero_h{0};
```
Comment (tangjj11) [inline clang/test/dpct/half_raw.cu:12]: how to handle `one_h+one_h`?

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -14606,3 +14605,43 @@ void CudaUuidRule::runRule(
...
+
+void HalfRawRule::runRule(
+    const ast_matchers::MatchFinder::MatchResult &Result) {
+  ExprAnalysis EA;
+  if (auto TL = getNodeAsType<TypeLoc>(Result, "halfRawType")) {
+    EA.analyze(*TL);
+    emplaceTransformation(EA.getReplacement());
+    EA.applyAllSubExprRepl();
+    return;
+  } else if (const InitListExpr *Init =
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: if return, need't else.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 3]
```diff
@@ -14606,3 +14605,43 @@ void CudaUuidRule::runRule(
...
+              MapNames::getClNamespace() + "half>(uint16_t(";
+    EA.analyze(Init->getInit(0));
+    OS << EA.getReplacedString();
+    OS << "))}";
+    OS.flush();
+    DpctGlobalInfo::getInstance().addReplacement(
+        std::make_shared<ExtReplacement>(SM, Init, Replacement, nullptr));
+    return;
+  } else {
+    return;
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: need't this branch?

**Review by tangjj11 — APPROVED** [review]

File: clang/test/dpct/half_raw.cu (hunk 1 above)
Comment (wenhuiNi) [reply clang/test/dpct/half_raw.cu:12]: __half_raw cannot be assigned， also __half_raw does not overload operator +

Comment (tangjj11) [reply clang/test/dpct/half_raw.cu:12]: it is ok, but we can use this: `one_h.x+one_h.x`.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 4]
```diff
@@ -14606,3 +14605,69 @@ void CudaUuidRule::runRule(
...
+    DpctGlobalInfo::getInstance().addReplacement(
+        std::make_shared<ExtReplacement>(
+            SM, ME->getOperatorLoc(),
+            SM.getFileOffset(ME->getEndLoc()) -
+                SM.getFileOffset(ME->getOperatorLoc()) + 1,
+            "", nullptr));
+    return;
+  }
+  if (auto DRE = getNodeAsType<DeclRefExpr>(Result, "halfRawExpr")) {
+    std::cout<<"enter this\n";
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: need delete

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 5]
```diff
@@ -14606,3 +14605,69 @@ void CudaUuidRule::runRule(
...
+    std::string Replacement;
+    llvm::raw_string_ostream OS(Replacement);
+    OS << "" + MapNames::getClNamespace() + "bit_cast<" +
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: Why need the first ""?

**Review by tangjj11 — CHANGES_REQUESTED** [review]

**Review by tangjj11 — APPROVED** [review]

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 6]
```diff
@@ -14689,3 +14689,68 @@ void CudaUuidRule::runRule(
...
+    if (ME->isArrow()) {
+      DpctGlobalInfo::getInstance().addReplacement(
+          std::make_shared<ExtReplacement>(SM, ME->getBeginLoc(), 0, "*",
+                                           nullptr));
+    }
+    DpctGlobalInfo::getInstance().addReplacement(
+        std::make_shared<ExtReplacement>(
+            SM, ME->getOperatorLoc(),
+            SM.getFileOffset(ME->getEndLoc()) -
+                SM.getFileOffset(ME->getOperatorLoc()) + 1,
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.cpp]: What's migration result of this(serials whitespace trailing `->`)
```
__half_raw->    member
```

Comment (wenhuiNi) [reply clang/lib/DPCT/ASTTraversal.cpp]: see lit: 
  // CHECK: *ptr = 0x3C00;
  ptr->x = 0x3C00;
  // CHECK: *ptr = 0x3C00;
  ptr ->x = 0x3C00;
  // CHECK: *ptr = 0x3C00;
  ptr-> x = 0x3C00;
  // CHECK: *ptr = 0x3C00;
  ptr -> x = 0x3C00;
  // CHECK: zero_h = 0x3C00;
  zero_h.x = 0x3C00;
  // CHECK: zero_h = 0x3C00;
  zero_h .x = 0x3C00;
  // CHECK: zero_h = 0x3C00;
  zero_h. x = 0x3C00;
  // CHECK: zero_h = 0x3C00;
  zero_h . x = 0x3C00;

**Review by yihwang-nv — APPROVED** [review]

File: clang/test/dpct/half_raw.cu [hunk 7]
```diff
@@ -0,0 +1,41 @@
+// UNSUPPORTED: cuda-8.0
+// UNSUPPORTED: v8.0
+// RUN: dpct --format-range=none -out-root %T/half_raw %s --cuda-include-path="%cuda-path/include" -- -std=c++14 -x cuda --cuda-host-only
+// RUN: FileCheck --input-file %T/half_raw/half_raw.dp.cpp --match-full-lines %s
+
+#include <cuda_fp16.h>
+#include <cuda_runtime.h>
+int main(){
```
Comment (tangjj11) [inline clang/test/dpct/half_raw.cu]: not format

**Review by ziranzha — APPROVED** [review]

**Pull request merged**